feat(AEP-5): add AEP around designing an API (#79)
aep-dev/aep.dev: aep/general/0005/aep.md

# Designing an API

> This AEP serves as a high-level guide on designing an API that adheres to the

Reviewer: ```suggestion
This AEP serves as a high-level guide to designing an API that adheres to the
```

This sounded a bit off to me, but I wasn't sure if it was just me, so I Googled it and found https://forum.wordreference.com/threads/guide-to-vs-on-vs-for.3645609/.

Separately, the structure of this sentence is a bit ambiguous:  "that adheres to the AEPs" could be describing "an API" or "This AEP".  Maybe something like "...designing an AEP-compliant API"?

Author: Sounds good! done.

AEPs.

## Process summary

> 1. enumerate the user case would like your API to satisfy.

Reviewer: ```suggestion
1. Enumerate the use cases you would like your API to satisfy.
```

I suggest capitalizing all of these.

Also, is "and add them to your API" on each of the next three lines necessary?

Author: I was adding the "add them to your API" since step 1 doesn't require them to do that. But we can try removing it.

1. identify resources and add them to your API.
1. identify standard operations and add them to your API.
1. identify custom operations and add them to your API.

## Enumerate use cases

> The first step in designing an API is understanding precisely what you would

Reviewer: Is this necessarily true?  Sometimes you might design an API and hope your users surprise you by building cool stuff with it.

Maybe "precisely which operations you would like your users to be able to perform"?

Author: Sure, I meant the latter - I kind of feel like "accomplish" and "perform" are synonymous here, but since you raised the comment I guess not. 

Applied your suggestion.

> like your user to be able to accomplish your API.

Reviewer: ```suggestion
like your users to be able to accomplish.
```

Author: updated to "user to perform", taking your other suggestion.

> Enumerate one or more use cases to clarify what precisel

Reviewer: I think you accidentally part of this sentence.

Author: Doh! Thanks and sorry.

For example:

- Create a VM.
- List all VMs owned by a company.
- List all VMs owned by a user.
- Restart a running VM.

> - Trade an in-game item with another player.

Reviewer: I appreciate the love for Roblox...but since this example doesn't fit with the others, maybe drop it?

(Or have two separate sets of examples, where the second one might include this and also things like "list items in a player's inventory", "propose a trade", "accept a trade", "find open trade offers containing an item", etc.)

Author: The list wasn't intended to be an example of a user designing an API themselves... just examples of use cases.

I see why it's confusing to lump them together, but I also think distinct sets of use cases will help consumers of the AEPs understand the scope is all APIs, not just cloud / games / etc.

I split them up, PTAL.

> Some best practices:

Reviewer: Nit:  A list of one thing probably doesn't need to be a list...

Author: I guess we could switch it later, but I'm hoping someone has more to add! 

Before I switch it back: any best practices we should include?

Reviewer: Maybe something about trying to plan for known or likely future use cases, rather than just the highest priority or MVP use cases?

Author: Added! PTAL.

- Attempt to define granular use cases that can be composed to
  satisfy more complex use cases.

## Identify resources

> Once your use cases are defined, consider how many of those can be accomplished
> by represented by an [API Resource Type][]: similar to object-oriented

Reviewer: ```suggestion
Once your use cases are defined, consider how many of those can
represented by an [API Resource Type][]: similar to object-oriented
```

I'm not sure if this was the original intent or not - just proofreading.

Author: Thanks! yeah it's the same outcome.

programming, these are use cases that can be represented by entities that are

> created, read, updated, and deleted.

Reviewer: I agree there are parallels between OOP and ROAD, but this reads like the main similarity is CRUD operations, and CRUD isn't really an OOP principle.

Author: yeah agreed, re-reading this is a false equivalency. removed.

Examples include:

- users

> - virtual machine

Reviewer: ```suggestion
- virtual machines
```

- load balancers
- services

> The core of the AEPs is resource-oriented design: this design paradigm allows

Reviewer: ```suggestion
The core concept described by the AEPs is resource-oriented design: this design paradigm allows
```

WDYT about this phrasing instead?

Author: describing to me implies that it's not very opinionated about the design paradigm, it's just explaining it. I was trying to say something that implies that one of the foundational concepts of the AEPs is resource-oriented design. 

I'll take the suggestion for now but I'd love to find a better wording.

for uniform standard operations that reduce the cognitive overhead in learning
about the operations and schemas exposed by your API.

Resources can relate to each other, either as a parent-child relationship

> defining ownership, or via references that defines the resources usage of

Reviewer: ```suggestion
defining ownership/scope, or via references that define the resource's usage of
```

> another resource in it's operations (e.g. a vm depending on a private network).

Reviewer: ```suggestion
another resource in its operations (e.g. a VM depending on a private network).
```

Author: I think your example is parent-child relationships, while the network interface was more a dependency.

But that makes me think we can structure this section better to enumerate these a bit. rewrote it, PTAL.

See the following AEPs to learn more about resource-oriented design:

> - [resource paths][AEP-122]

Reviewer: We also want 121 and 123 here, right?  I know we haven't adopted them yet, but maybe we just have some future-compatible and temporarily broken links here?  Otherwise we have to remember to update this, and realistically we probably won't.

Author: sure, added.

## Identify standard operations

Once the resources are defined, identify one or more standard methods for each
of those resources. Standard methods operate on the lifecycle of a resource
lifecycle: namely, they create, read, update, delete, and list resources.

Ideally all 5 standard methods should be exposed for every resource.

See the following AEPs to learn more about the standard operations:

- [AEP-131][]

## Identify custom operations

> To accomplish some of the user journeys, resources may need to support more
> than just operations to create, read, update, delete, and list them. Some

Reviewer: ```suggestion
To accomplish some of the user journeys, resources may need to support operations
other than Create, Update, Delete, Get, and List. Some
```

I made a couple small changes here, WDYT?  Maybe we also link each CUDGL operation to the relevant AEP?

Author: taking the wording suggestion!

I'll add the links too, but I think we *really* need to speed up adding AEPs if we do this - it's very unprofessional for a project to be littered with broken links everywhere, I think it'll impact the legitimacy of our AEPs if I point folks over to them.

examples include:

> - restarting a virtual machine.
> - cleaning a disk.

Reviewer: ```suggestion
- restarting a virtual machine
- cleaning a disk
```

Reviewer: What is "cleaning" a disk?  Maybe "wiping"?

Also, as a nit, can we add a third example here?  Lists with at least three elements feel better IMO. :)

Author: done.

# Designing an API

This AEP serves as a high-level guide to designing an AEP-compliant API. AEPs.

## Process summary

1. Enumerate the use cases you would like your API to satisfy.
1. Identify resources.
1. Identify standard operations.
1. Identify custom operations.

## Enumerate use cases

The first step in designing an API is understanding precisely what operations

> you would like your user to perform. Enumerate these operations, attempting to

Reviewer: ```suggestion
you would like your user to be able to perform. Enumerate these operations, attempting to
```

  complex use cases.

## Identify resources

Once your use cases are defined, consider how many of those can be represented
by an [API Resource Type][]: entities that are created, read, updated, and
deleted.

Examples include:

- users
- virtual machines
- load balancers
- services

One of the core concepts described by the AEPs is resource-oriented design:
this design paradigm allows for uniform standard operations that reduce the
cognitive overhead in learning about the operations and schemas exposed by your
API.

Resources can relate to each other. For example:

- A parent-child relationship defining ownership/scope (A user who created a
  VM).
- A resource dependency, where one resource depends on another to function (A
  VM depending on a disk).

See the following AEPs to learn more about resource-oriented design:

- [resource-oriented design][AEP-121]
- [resource paths][AEP-122]
- [resource types][AEP-123]

## Identify standard operations

Once the resources are defined, identify one or more standard methods for each
of those resources. Standard methods operate on the lifecycle of a resource
lifecycle: namely, they create, read, update, delete, and list resources.

> Ideally all 5 standard methods should be exposed for every resource.

Reviewer: ```suggestion
Ideally all five standard methods should be exposed for every resource.
```

Nit:  https://www.grammarly.com/blog/when-to-spell-out-numbers/

> See the following AEPs to learn more about the standard operations:

Reviewer: ```suggestion
See the following AEPs to learn more about the standard methods:
```

We should probably be consistent about whether we say "standard methods" or "standard operations".

It's a bigger issue than this PR; aep.dev has a "standard methods" category containing a "custom operations" AEP.
